[User guides] Add user guides for DeepSpeed and Accelerate (#38513)
ray-project/ray: doc/source/train/huggingface-accelerate.rst

and call `Accelerator.prepare() <https://huggingface.co/docs/accelerate/main/en/package_reference/accelerator#accelerate.Accelerator.prepare>`_ 
as usual to prepare everything for distributed training.

You can expect the final code to look like this:

.. code-block:: python

    from accelerate import Accelerator

    def train_func(config):
        # Instantiate the accelerator
        accelerator = Accelerator(...)

        model = ...
        optimizer = ...
        train_dataloader = ...
        eval_dataloader = ...
        lr_scheduler = ...

        # Prepare everything for distributed training
        (
            model,
            optimizer,
            train_dataloader,
            eval_dataloader,
            lr_scheduler,
        ) = accelerator.prepare(
            model, optimizer, train_dataloader, eval_dataloader, lr_scheduler
        )

        # Start training
        ...
    
    from ray.train.torch import TorchTrainer
    from ray.train import ScalingConfig

    trainer = TorchTrainer(
        train_func,
        scaling_config=ScalingConfig(...),
        ...
    )
    trainer.fit()

Next, check these end-to-end examples below for more details:

> - Basic: :ref:`Accelerate Distributed Data Parallel Training with Ray Train <accelerate_example>`

Reviewer: Should we `literalinclude` this one?

You may find these user guides helpful:

- :ref:`Configuring Scale and GPUs <train_scaling_config>`
- :ref:`Configuration and Persistent Storage <train-run-config>`
- :ref:`Saving and Loading Checkpoints <train-checkpointing>`
- :ref:`How to use Ray Data with Ray Train <data-ingest-torch>`

`AccelerateTrainer` Migration Guide 
-----------------------------------

Before Ray 2.7, Ray Train's :class:`AccelerateTrainer <ray.train.huggingface.AccelerateTrainer>` API was the 
recommended way to run Accelerate code. As a subclass of :class:`TorchTrainer <ray.train.torch.TorchTrainer>`, 
``AccelerateTrainer`` helps users apply the configuration file generated with ``accelerate config`` to all workers. 
Aside from that, the functionality of ``AccelerateTrainer`` is identical to :class:`TorchTrainer <ray.train.torch.TorchTrainer>`.

However, users can easily configure Accelerate within their training function, so there is no need to maintain a separate Ray Trainer API.
To provide a more simplified and flexible interface, we will deprecate ``AccelerateTrainer`` in Ray 2.8, and recommend running your 
Accelerate code with :class:`TorchTrainer <ray.train.torch.TorchTrainer>`. 

> How to setup Accelerate config

Reviewer: Organizationally I think it would be nice to put this near the top, maybe before the examples. This information is helpful for the general audience, and not just for people migrating from `AccelerateTrainer`

    like :meth:`~ray.train.torch.prepare_model` or :meth:`~ray.train.torch.prepare_data_loader` in your training funciton. Instead, 
    you can just instantiate an `Accelerator <https://huggingface.co/docs/accelerate/main/en/package_reference/accelerator#accelerate.Accelerator>`_ 
    and call `Accelerator.prepare() <https://huggingface.co/docs/accelerate/main/en/package_reference/accelerator#accelerate.Accelerator.prepare>`_ 
    as usual to prepare everything for distributed training.

How to setup Accelerate config
------------------------------

Hugging Face Accelerate provides two approaches to configure your run:

1. Use `accelerate config` to generate a configuration YAML file, and use `accelerate launch $YOUR_PY_SCRIPT --config_file=$YOUR_CONFIG_YAML` to launch via CLI. 
2. Set configurations through the `accelerate.Accelerator <https://huggingface.co/docs/accelerate/main/en/package_reference/accelerator#accelerate.Accelerator>`_ object in your Python script, then run your script with a custom launcher.

In Ray Train, you should always use the second approach, where Ray TorchTrainer acts as a distributed launcher and automatically sets up 

> the Torch distributed environment and run thr training function on all workers.

Author: Make sense. I can remove the enumerate points, and focus on discussing the second approach. And add an notice for accelerate config at the end of this section?

> the Torch distributed environment and run thr training function on all workers.

Reviewer: Yeah, no need to make it stand out though. I think the point to call out is that you don't need the CLI command since TorchTrainer configures the launching part for you.

                    ...,
                    fsdp_plugin=fsdp_plugin,
                )

                # Start training
                ...

            from ray.train.torch import TorchTrainer
            from ray.train import ScalingConfig

            trainer = TorchTrainer(
                train_func,
                scaling_config=ScalingConfig(...),
                ...
            )
            trainer.fit()

Next, check these end-to-end examples below for more details:

.. dropdown:: End-to-end Code Example

    .. literalinclude:: /../../python/ray/train/examples/accelerate/accelerate_torch_trainer.py
        :language: python

.. seealso::

    If you're looking for more advanced use cases, check out this Llama-2 fine-tuning example: 
    
    - `Fine-tuning Llama-2 series models with Deepspeed, Accelerate, and Ray Train. <https://github.com/ray-project/ray/tree/master/doc/source/templates/04_finetuning_llms_with_deepspeed>`_

You may also find these user guides helpful:

- :ref:`Configuring Scale and GPUs <train_scaling_config>`
- :ref:`Configuration and Persistent Storage <train-run-config>`
- :ref:`Saving and Loading Checkpoints <train-checkpointing>`

> - :ref:`How to use Ray Data with Ray Train <data-ingest-torch>`

Author: I think these are still useful, since the accelerate users still need to find a way to configure `TorchTrainer`.

`AccelerateTrainer` Migration Guide 
-----------------------------------

> Before Ray 2.7, Ray Train's :class:`AccelerateTrainer <ray.train.huggingface.AccelerateTrainer>` API was the 
> recommended way to run Accelerate code. As a subclass of :class:`TorchTrainer <ray.train.torch.TorchTrainer>`,  
> AccelerateTrainer helps users apply the configuration file generated with ``accelerate config`` to all workers. 

Reviewer: ```suggestion
Before Ray 2.7, Ray Train's :class:`AccelerateTrainer <ray.train.huggingface.AccelerateTrainer>` API was the 
recommended way to run Accelerate code. As a subclass of :class:`TorchTrainer <ray.train.torch.TorchTrainer>`,  
the AccelerateTrainer takes in a configuration file generated by ``accelerate config`` and applies it to all workers. 
```

Aside from that, the functionality of ``AccelerateTrainer`` is identical to ``TorchTrainer``.

> However, users can easily configure Accelerate within their training function, so there is no need to maintain a separate Ray Trainer API.
> To provide a more simplified and flexible interface, we will deprecate ``AccelerateTrainer`` in Ray 2.8, and recommend running your 
> Accelerate code with ``TorchTrainer``. 

Reviewer: ```suggestion
However, this caused confusion around whether this was the *only* way to run Accelerate code. 
Because the full Accelerate functionality can be expressed with the ``Accelerator`` and ``TorchTrainer`` combination, the ``AccelerateTrainer`` will be deprecated in Ray 2.8, 
and it is recommend to run your  Accelerate code directly with ``TorchTrainer``. 
```
